### DISCUSSION TITLE
[Version 2] Task List

### DISCUSSION THREAD

**Discussion by magicbug** [magicbug/Cloudlog#1283] [General]
Progress with version 2.0 is coming along if you have been watching the station_logbook branch, a big thankyou to @AndreasK79 and @Werzi2001 for their help, however, there are lots of bits that still need to be completed and I wanted to keep people in the loop to get V2.0 ready for launching the items below will need completing, once this is done and tested hopefully with PHP 8.0 support it will be merged into the Master branch and another branch "Development" will be created to hold future development work.

Things will change after this and Development will become the "BETA" branch so to speak, everything will be done on there then every week or so merged into master, this should reduce bugs in the master branch.

There's been a big change in the roles there's now just two these are "Admin" and "Operator" this allows a single instance to support more than one user (finally!)

I'm sure there are lots of "new features" everyone would like but this is the bare minimum at the moment, new features will be handled separately when the following works.

# Install

- [ ] Improved with a checklist on first login, with system to create a new user

# User Accounts

- [x] Password reset system

# Dashboard

- [x] Countries Breakdown needs to reflect ALL QSOS from the selected Logbook
- [x]  QSL Cards needs to reflect ALL QSOs from the selected logbook

# QSO

- [x] Previous Contacts must only show QSOs from the selected logbook
- [x] Previous Contacts after QSO entry must only show PAST QSOs within the logged in users station locations
- [x] QSL cards can only be shown to the user who uploaded them

## QSL Cards

- [x] Just need to make sure only see your own cards

# Analytics

- [x] Suffers the same implode error as the Awards items when zero QSOs are found.

## Statistics

- [x] Generally would love to move away from Google Charts however low priority.
- [x] Need to handle zero QSOs better lots of errors.
- [x] Hide "Satellites" unless Satellite QSOs are found
- [x] Suffers same implode error like the Awards items

# Awards

- [x] General if no QSOs are found handle errors better rather than implode errors

## DXCC Award

- [X] By default, Deleted DXCCs need to be hidden as its only relevant to a small group

## IOTA Award

- [X] Deleted IOTA Islands hidden by default rather than shown

# Search

- [X] Only show results relevant to the logged-in user who have access to the logbooks and qsos available.

# User Dropdown

## Station Logbook

- [x] If no logbooks have been created when making a new one set it to active.

## Station Location

- [x] When creating a location it must set the user_id

## ADIF Import / Export

- [x] Change "Select Station Profile" to "Select Station Location"
- [x] In the Select Station Profile only show locations for the logged-in user for both import and export and lotw

## Print Requested QSLs

- [x] Only show "logged in users" station locations
- [x] Only show logged in users QSOs

## KML Export

- [x] Make sure it only exports for the current "Logbook" only

## DX Atlas Gridsquare Export

- [x] Make sure it only deals with the logged in users station locations

## LoTW

- [x] Only show certs for logged in user
- [x] Manual Sync only triggers for that one user (still needs to do everyone as cron), this could be hidden and only shown to admins then allow "operators" to sync per location inside the Station Locations area.
- [x] Import - Needs to only show data for that user.

## eQSL
- [x] Need to improve the error displayed if no eQSL Nicks have been set
- [x] Only upload / import for the "logged in user no one else"

## Clublog

- [x] Script needs to be rewritten to not require a username in the cron and to run across all users.

## QRZ

- [x] Web UI needs to only show items for the logged-in user

## API Keys

- [x] Operator Role needs access
- [x] Only show Keys for the logged-in user

## Hardware Interfaces
- [x] Operator Role needs access
- [x] Only show hardware interfaces for the logged-in user.

# Public Dashboard

- [x] This needs to be coded as per the discussion so that it's available.

**Comment by dg9vh** [comment]
Print Requested QSLs
Only show "logged in users" station locations
Only show logged in users QSOs

To this point I would suggest that you can define one user as a
"logbook-master-account" that can show and print all log-entries to solve
following use-case that would be very common on club-stations:

Here many individual operators will use a common callsign assigned to one
person as a club-callsign (for example: DLØNK). Now the callsign-keeper
should be able to print all QSOs logged for this callsign but done from
several operators that have individual logins for the logbook.

Just my 2 cent :-)

73 de Kim
DG9VH

**Reply by magicbug** [reply]
Logbooks do have a Master user_id of who created them so thats no issue to implement

**Comment by dg9vh** [comment]
Would be perfect :-)

Am Sa., 13. Nov. 2021 um 22:45 Uhr schrieb Peter Goodhall <
***@***.***>:

> Logbooks do have a Master user_id of who created them so thats no issue to
> implement
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/magicbug/Cloudlog/discussions/1283#discussioncomment-1637397>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKN5OU7TCVYJMZOGATJE4TUL3L6FANCNFSM5H7B62BA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Amateurfunk ist das, was uns verbindet! Bleiben wir in Kontakt!
Ortsverband Q03, Neunkirchen

**Comment by AndreasK79** [comment]
An extensive list indeed. I didn't realize there were so many, but they do make sense.

To make things easier, I would suggest merging master into station_logbooks to get it up to date.
Also merging the PHP8 PR would be nice. By doing this, we would avoid too many merge conflicts, since fixing this list would make changes to many of the same files.

**Reply by magicbug** [reply]
Do you want to deal with that @AndreasK79 merge it in and then we can deal with any fall out

**Reply by AndreasK79** [reply]
@magicbug Yes, I can do that. I did it yesterday on my own account.

**Reply by magicbug** [reply]
@AndreasK79 Errm think Station Logbooks just got merged into Master instead of the other way

**Reply by AndreasK79** [reply]
@magicbug you are right. I picked the wrong option when fixing merge conflict. But I think it should be fixed by reverting the one that says merge.

**Reply by magicbug** [reply]
@AndreasK79 Yep on it

**Reply by AndreasK79** [reply]
Hmm, doesn't seem like that did it. Can we swap out this master with my master? That one is the same as before the merge.

**Reply by magicbug** [reply]
Reverted and seems OK with the code "as is"

**Reply by AndreasK79** [reply]
Great. Sorry about that. I'm trying to be as careful as I can, but sometimes I stumble :(

**Reply by magicbug** [reply]
I'm pretty certain i've left the PHP8 changes in the "master branch" but doesn't seem to have broken local dev copy soo fingers crossed

**Reply by magicbug** [reply]
I think https://github.com/magicbug/Cloudlog/pull/1285 needs rejected as it removes most of the "station logbook" code sadly

**Reply by magicbug** [reply]
Looks like things are stable again, back to coding ;-)

**Reply by AndreasK79** [reply]
Good. This thing is giving me a headache. The only thing I'm wondering about now, is how we are going to get the changes from master over in station_logbooks.

**Reply by magicbug** [reply]
I'll cherry pick them over later :)

**Reply by AndreasK79** [reply]
Great, thanks. One suggestion would be to merge in my cloudlog8-branch. That contains everything from master+station_logbooks+php8, but without my screwup.

**Comment by AndreasK79** [comment]
A few comments and ideas regarding install (and first setup).

- When you enter Cloudlog the first time, it only says config-file not found. If you want avoid this, Codeigniter must be changed. My suggestion would be to change common.php like this:
`				echo 'The configuration file does not exist.';
				if (file_exists('install/')) {
					header('Location: install');
				}`
At least the user would get a redirect to run the install.

- When getting to the install, a link to the wiki would be nice, in case the user doesn't know what to do. Something like this perhaps:

![image](https://user-images.githubusercontent.com/6977712/141675618-ce4c169a-d79b-4b5e-827b-db558dd55c94.png)

- When the install is done, redirect to the main Cloudlog url (if everything is successful that is).
- Present a list of things the user need to do when first logging in (update country files, make station logbook, make station location).
- Somehow merge station logbooks and station locations into one view. I would think many people will find the setup confusing then trying out V2. Add text in the view to explain what needs to be done and why. This will help the users help themselves.

This is what comes to mind when installing Cloudlog from scratch, and have no idea how everything is connected together.

**Reply by magicbug** [reply]
Seems like some decent suggestions, I'm not sure how best to merge Logbooks & Locations, perhaps "Logbooks" become the core item then inside that we have a popup for creating locations?

**Comment by Werzi2001** [comment]
@magicbug Maybe it's a good idea to mark (strike through?) done todos in your list (in order to avoid confusion). I created pull requests for the dashboard, DXCC and IOTA tasks. Chosen some easy stuff so you guys can focus on the more important things :)

**Reply by Werzi2001** [reply]
Oh just saw you already did. Nevermind.

**Reply by magicbug** [reply]
Really appreciated! @Werzi2001 helps hugely!

**Comment by AndreasK79** [comment]
@magicbug What kind of implode error do you get on analytics and awards? I don't see them here. I mostly get "Nothing found".

**Reply by magicbug** [reply]
This is the implode error when zero qsos found

![Screenshot from 2021-11-14 12-53-32](https://user-images.githubusercontent.com/84308/141681896-4a27317b-f2eb-4605-bfac-f82e23aabcf3.png)

Also getting

![Screenshot from 2021-11-14 12-53-41](https://user-images.githubusercontent.com/84308/141681912-34b1618d-f9af-4ba8-9521-bc368513f02a.png)

On the Stats Graphs page since PHP8 merge

**Reply by magicbug** [reply]
logbooks_model error only seems to happen on user accounts with no qsos

**Reply by AndreasK79** [reply]
Thanks Peter. My local install doesn't display those errors. 

That means everything that uses a band-dropdown is affected (awards, analytics, kml export, dxatlas export, custom maps). You can put me on fixing the implode message.

**Reply by magicbug** [reply]
Okay! nearly finished fixing the  logbook_model errors so I'll leave the impode to you :D

**Reply by AndreasK79** [reply]
Super. I finally found the reason for not getting the error. Debugging in VSCode sure does help. 

Fixing the implode error seems easy enough, but I get a lot of errors when I have no QSOs, locations or logbooks. Should these errors be fixed as well?

**Reply by magicbug** [reply]
If we can yes, or just hide the awards until QSO records have been found ?

**Reply by AndreasK79** [reply]
The easy way would be to hide, but we should really have some more checks, so I'll fix it.

**Reply by AndreasK79** [reply]
Fixed all awards. Almost fixed all analytics. Only gridsquares and statistics left.

**Reply by AndreasK79** [reply]
PR pushed. Gridsquares and statistics also fixed.

**Comment by magicbug** [comment]
Need some discussion on Clublog uploading at the moment needs a cloudlog "username" passed that holds the clublog login, I'm thinking going forward this isn't the best solution and with multi logbooks and users would it be better suited to move the clublog logins to either Logbooks or Locations.

This could also be done for LoTW so that different logins can be used for different callsigns especially if its club vs home call 

Feedback welcome!

**Reply by AndreasK79** [reply]
It should be moved to locations. As the calls may or may not be under the same account on Club Log.

I think the same should be done for eQSL as @Werzi2001 suggested here #1147.

**Reply by magicbug** [reply]
Seems sane, probably should add a separate table to store login details and then assign it to the profile, I'm thinking to allow quickly adding.

third_party_logins
ID | User_ID | Provider | User | Pass | Date Modified

Ideally encrypting the important parts

**Reply by timkirby** [reply]
> Need some discussion on Clublog uploading at the moment needs a cloudlog "username" passed that holds the clublog login, I'm thinking going forward this isn't the best solution and with multi logbooks and users would it be better suited to move the clublog logins to either Logbooks or Locations.
> 
> This could also be done for LoTW so that different logins can be used for different callsigns especially if its club vs home call
> 
> Feedback welcome!

I'd find this approach very helpful for LOTW too at the moment, although I can upload my GW4MM QSOs, I can't download and sync them because a different user id is used for my two callsigns on the LOTW site...

**Comment by AndreasK79** [comment]
@magicbug API keys only show keys based on user_id. I added that when I added user_id to the table.

 
> // GET API Keys
    function keys() {
		$this->db->where('user_id', $this->session->userdata('user_id'));
    	return $this->db->get('api');
    }

**Comment by AndreasK79** [comment]
@magicbug regarding qslcards, it can be checked. @Werzi2001 added this in a PR. It checks on station_id for the logged in user, both on loading page and searching for additional QSOs.

**Reply by AndreasK79** [reply]
@magicbug the same with QSO and qslcards. @Werzi2001 added a check there too :)

**Comment by AndreasK79** [comment]
@magicbug just checked the code for hardware interfaces. models/cat/status only loads result from cat-table with logged in user.

**Comment by magicbug** [comment]
Hi All,

Just to say thanks @Werzi2001 and @AndreasK79 for the PRs most are now merged, finally have some free time which makes a nice change :) shall try to work through my list in the coming days :)

**Comment by magicbug** [comment]
Password Reset is now in, it uses default "SendMail" at the moment but I'm working on adding email settings to global configuration area next.

![Screenshot from 2022-01-18 17-56-18](https://user-images.githubusercontent.com/84308/149992494-180ed070-140e-4643-b2e3-30783b63c536.png)

Just need to save config items then wire it all in.

**Reply by magicbug** [reply]
This is now implemented, this should hopefully allow more email items to be sent via Cloudlog in the long run maybe.

**Comment by AndreasK79** [comment]
@magicbug I checked the code regarding LoTW. It only loads the certs for the logged in userid, so that's one you can check as done.

**Reply by magicbug** [reply]
Checked off.

**Comment by magicbug** [comment]
Just so I don't forget when using /api/qso I have added a switch to the import call that enables API mode which means it doesn't check the user against the station profile for now.

**Comment by magicbug** [comment]
Took a bit of time out of new features and looked over the dashboard code to see what could be improved before reusing some of the code for the public display managed to get it down from 40+ SQL queries down to 16 with an average page speed increase of roughly 60%.

**Reply by AndreasK79** [reply]
That's great. It was sure noticeable when loading the page :)

**Reply by magicbug** [reply]
Was worth the time :-) back to coding the public display just thought I might as well tidy it up.

**Comment by AndreasK79** [comment]
@magicbug where are we regarding this list?

I can take a look at the Clublog script.

**Comment by AndreasK79** [comment]
> Generally would love to move away from Google Charts however low priority:

Work in progress:
![image](https://user-images.githubusercontent.com/6977712/190878148-2b2e028a-02da-4e91-9089-542db8660531.png)
![image](https://user-images.githubusercontent.com/6977712/190878155-355fe9e0-2e0a-41e1-bed0-8de603b32825.png)
![image](https://user-images.githubusercontent.com/6977712/190878165-fb010b92-2714-47ab-896c-36bd6fe387bd.png)
![image](https://user-images.githubusercontent.com/6977712/190878207-1786024d-0de8-4b7c-8acb-e14a14ade331.png)

Still got some tweaking to do, but it's pretty close.

**Comment by AndreasK79** [comment]
Only eQSL and setup checklist is left in the list on top. Seems to me we are close.

**Reply by magicbug** [reply]
Yep! very close, Setup Checklist I was thinking more a "Welcome to Cloudlog version 2.0" style page that gets shown and walks people through the changes.

There is some eQSL PRs but I haven't looked at them to see if they address anything

**Reply by AndreasK79** [reply]
That's a lot of changes... :)
